kernel_test.go

```diff
@@ -15,6 +16,249 @@ const (
+	}
+
+	// wait for a second to give the tcp connection time to complete to avoid missing the early pub messages
+	time.Sleep(1 * time.Second)
+
+	return testJupyterClient{shell, iopub}, func() {
+		if err := shell.Close(); err != nil {
+			t.Fatal("shell.Close", err)
+		}
+		if err = iopub.Close(); err != nil {
+			t.Fatal("iopub.Close", err)
+		}
+	}
+}
+
+// sendShellRequest sends a message to the kernel over the shell channel. Upon error, sendShellRequest
+// will Fail the test.
+func (client *testJupyterClient) sendShellRequest(t *testing.T, request ComposedMsg) {
+	t.Helper()
+
+	if _, err := client.shellSocket.Send("<IDS|MSG>", zmq.SNDMORE); err != nil {
+		t.Fatal("shellSocket.Send:", err)
+	}
+
+	reqMsgParts, err := request.ToWireMsg([]byte(connectionKey))
+	if err != nil {
+		t.Fatal("request.ToWireMsg:", err)
+	}
+
+	if _, err = client.shellSocket.SendMessage(reqMsgParts); err != nil {
+		t.Fatal("shellSocket.SendMessage:", err)
+	}
+}
+
+// recvShellReply tries to read a reply message from the shell channel. It will timeout after the given
+// timeout delay. Upon error or timeout, recvShellReply will Fail the test.
+func (client *testJupyterClient) recvShellReply(t *testing.T, timeout time.Duration) (reply ComposedMsg) {
+	t.Helper()
+
+	ch := make(chan ComposedMsg)
+
+	go func() {
+		repMsgParts, err := client.shellSocket.RecvMessageBytes(0)
+		if err != nil {
+			t.Fatal("Shell socket RecvMessageBytes:", err)
+		}
+
+		msgParsed, _, err := WireMsgToComposedMsg(repMsgParts, []byte(connectionKey))
+		if err != nil {
+			t.Fatal("Could not parse wire message:", err)
+		}
+
+		ch <- msgParsed
+	}()
+
+	select {
+	case reply = <-ch:
+	case <-time.After(timeout):
+		t.Fatal("recvShellReply timed out")
+	}
+
+	return
+}
+
+// recvIOSub tries to read a published message from the IOPub channel. It will timeout after the given
+// timeout delay. Upon error or timeout, recvIOSub will Fail the test.
+func (client *testJupyterClient) recvIOSub(t *testing.T, timeout time.Duration) (sub ComposedMsg) {
+	t.Helper()
+
+	ch := make(chan ComposedMsg)
+
+	go func() {
+		repMsgParts, err := client.ioSocket.RecvMessageBytes(0)
+		if err != nil {
+			t.Fatal("IOPub socket RecvMessageBytes:", err)
+		}
+
+		msgParsed, _, err := WireMsgToComposedMsg(repMsgParts, []byte(connectionKey))
+		if err != nil {
+			t.Fatal("Could not parse wire message:", err)
+		}
+
+		ch <- msgParsed
+	}()
+
+	select {
+	case sub = <-ch:
+	case <-time.After(timeout):
+		t.Fatal("recvIOSub timed out")
+	}
+
+	return
+}
+
+// request preforms a request and awaits a reply on the shell channel. Additionally all messages on the IOPub channel
+// between the opening 'busy' messages and closing 'idle' message are captured and returned. The request will timeout
+// after the given timeout delay. Upon error or timeout, request will Fail the test.
+func (client *testJupyterClient) request(t *testing.T, request ComposedMsg, timeout time.Duration) (reply ComposedMsg, pub []ComposedMsg) {
```

Comment:
Maybe or more instructive name for this function? `sendJupyterRequest` or something?